From 17ea5a0c7b47ce99e3f8103169aabc3343033b49 Mon Sep 17 00:00:00 2001
From: WRS Support <support@windriver.com>
Date: Fri, 2 Oct 2009 16:12:13 -0400
Subject: [PATCH] hypervisor: DMA changes for hypervisor/guest

This represents changes to existing DMA files in
order to support linux as a guest OS on WR hypervisor.

---
 kernel/dma-coherent.c        |   23 ++++++++++++++++++++++-

diff --git a/kernel/dma-coherent.c b/kernel/dma-coherent.c
index c1d4d5b..c049dc8 100644
--- a/kernel/dma-coherent.c
+++ b/kernel/dma-coherent.c
@@ -5,6 +5,10 @@
 #include <linux/kernel.h>
 #include <linux/dma-mapping.h>
 
+#ifdef CONFIG_WRHV
+#include <vbi/interface.h>
+#endif
+
 struct dma_coherent_mem {
 	void		*virt_base;
 	u32		device_base;
@@ -116,8 +120,25 @@ int dma_alloc_from_coherent(struct device *dev, ssize_t size,
 		int page = bitmap_find_free_region(mem->bitmap, mem->size,
 						     order);
 		if (page >= 0) {
-			*dma_handle = mem->device_base + (page << PAGE_SHIFT);
 			*ret = mem->virt_base + (page << PAGE_SHIFT);
+#ifdef CONFIG_WRHV
+			if (paravirt_enabled()) {
+				u64 paddr;
+				dma_addr_t addr = mem->device_base
+						  + (page << PAGE_SHIFT);
+
+				if (vbi_get_guest_dma_addr((void *)addr,
+							&paddr) == 0)
+					*dma_handle = (dma_addr_t)paddr;
+				else {
+					bitmap_release_region(mem->bitmap,
+							      mem->size, order);
+					return 0;
+				}
+			} else
+#endif
+			*dma_handle = mem->device_base + (page << PAGE_SHIFT);
+
 			memset(*ret, 0, size);
 		} else if (mem->flags & DMA_MEMORY_EXCLUSIVE)
 			*ret = NULL;
-- 
1.6.5.2

